core/services/pipeline/orm.go

```diff
@@ -0,0 +1,194 @@
+	AwaitRun(ctx context.Context, runID int64) error
+	NotifyCompletion(pipelineRunID int64) error
+	ResultsForRun(runID int64) ([]Result, error)
+
+	// BridgeTask
+	FindBridge(name models.TaskType) (models.BridgeType, error)
+}
+
+type orm struct {
+	db *gorm.DB
+}
+
+var _ ORM = (*orm)(nil)
+
+func NewORM(db *gorm.DB) *orm {
+	return &orm{db}
+}
+
+// CreateRun adds a Run record to the DB, and one TaskRun
+// per TaskSpec associated with the given Spec.  Processing of the
+// TaskRuns is maximally parallelized across all of the Chainlink nodes in the
+// cluster.
+func (o *orm) CreateRun(jobSpecID int32) (int64, error) {
+	var runID int64
+	err := ormpkg.GormTransaction(o.db, func(tx *gorm.DB) error {
+		// Fetch the spec ID
+		var spec Spec
+		err := o.db.
+			Where("job_spec_id = ?", jobSpecID).
+			First(&spec).
+			Error
+		if err != nil {
+			return err
+		}
+
+		// Create the job run
+		run := Run{
+			SpecID:    spec.ID,
+			CreatedAt: time.Now(),
+		}
+
+		err = tx.Create(&run).Error
+		if err != nil {
+			return errors.Wrap(err, "could not create pipeline run")
+		}
+
+		runID = run.ID
+
+		// Create the task runs
+		err = tx.Exec(`
+            INSERT INTO pipeline_task_runs (
+                pipeline_run_id, pipeline_task_spec_id, created_at
+            )
+            SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, NOW() AS created_at
+            FROM pipeline_task_specs
+            WHERE pipeline_spec_id = ?
+        `, run.ID, run.SpecID).Error
+		return errors.Wrap(err, "could not create pipeline task runs")
+	})
+	return runID, err
+}
+
+// WithNextUnclaimedTaskRun chooses any arbitrary incomplete TaskRun from the DB
+// whose parent TaskRuns have already been processed.
+func (o *orm) WithNextUnclaimedTaskRun(fn func(ptRun TaskRun, predecessors []TaskRun) Result) error {
+	return ormpkg.GormTransaction(o.db, func(tx *gorm.DB) error {
+		var ptRun TaskRun
+		var predecessors []TaskRun
+
+		// NOTE: This could conceivably be done in pure SQL and made marginally more efficient by preloading with joins
+
+		// Find the next unlocked, unfinished pipeline_task_run with no uncompleted predecessors
+		err := tx.Table("pipeline_task_runs AS successor").
+			Set("gorm:query_option", "FOR UPDATE OF successor SKIP LOCKED"). // Row-level lock
+			Joins("LEFT JOIN pipeline_task_runs AS predecessors ON successor.id = predecessor.successor_id AND predecessors.finished_at IS NULL").
+			Where("predecessors.id IS NULL").
+			Where("successor.finished_at IS NULL").
+			Preload("TaskSpec").
+			Order("id ASC").
+			First(&ptRun).
+			Error
+		if err != nil {
+			return errors.Wrap(err, "error finding next task run")
+		}
+
+		// Find all the predecessors
+		err = tx.Where("successor_id = ?", ptRun.ID).Find(&predecessors).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding task run predecessors")
+		}
+
+		// Call the callback and convert its output to a format appropriate for the DB
+		result := fn(ptRun, predecessors)
```

Comment:
can the run error?

Reply:
The error is contained in the `Result` struct.  Processing should continue on because of tolerant adapters like `median`